src/main/java/frc/robot/subsystems/climb/Climb.java

```diff
@@ -0,0 +1,220 @@
+
+    private Double _climbSetpointLeft  = null;
+    private Double _climbSetpointRight = null;
+
+    private double _climbMinExtension  = Constants.Climb.MIN_EXTENSION; // TODO: tune value
+    private double _climbMaxExtension  = Constants.Climb.MAX_EXTENSION; // TODO: tune value
+
+    public Climb(ClimbIO io)
+    {
+        _io = io;
+
+        _climbFeedbackLeft  = new PIDController(0, 0, 0); // TODO: tune values
+        _climbFeedbackRight = new PIDController(0, 0, 0);
+
+        _tiltPID  = new PIDController(0, 0, 0); // TODO: tune values
+        _leftPID  = new PIDController(0, 0, 0);
+        _rightPID = new PIDController(0, 0, 0);
+
+        _gyro = new AHRS(Port.kMXP);
+
+        // Takes the desired position of both arms, and takes the needed adjustment calculated from the tilt to set the position - setpoint(?) for both arms
+        _leftHeight  = _desiredHeight + _heightAdjustment;
+        _rightHeight = _desiredHeight - _heightAdjustment;
+    }
+
+    @Override
+    public void periodic()
+    {
+        _io.updateInputs(_inputs);
+        Logger.processInputs("Climb", _inputs);
+
+        if (_climbSetpointLeft != null)
+        {
+            _io.setVoltageLeft(_climbFeedbackLeft.calculate(getExtensionLeft(), _climbSetpointLeft));
+        }
+
+        if (_climbSetpointRight != null)
+        {
+            _io.setVoltageRight(_climbFeedbackRight.calculate(getExtensionRight(), _climbSetpointRight));
+        }
+
+        _heightAdjustment = _tiltPID.calculate(getCurrentTilt(), 0);
+
+        // Gets the current position of the left and the right arms, and sets it to the desired position based on the tilt input, also applies this voltage
+        if(_leftHeight != null)
+        {
+            _io.setAlgorithmVoltageLeft(_leftPID.calculate(getExtensionLeft(), _leftHeight));
+        }
+
+        if(_rightHeight != null)
+        {
+            _io.setAlgorithmVoltageRight(_rightPID.calculate(getExtensionRight(), _rightHeight));
+        }
+    }
+
+    public void stop()
+    {
+        _io.setVoltageLeft(0.0);
+        _io.setVoltageRight(0.0);
+
+        _climbSetpointLeft  = null;
+        _climbSetpointRight = null;
+
+        setLockState(true);
+    }
+
+    public void setLockState(boolean enabled)
+    {
+        _io.setLockStateLeft(enabled, _inputs);
+        _io.setLockStateRight(enabled, _inputs);
+    }
+
+    public void setSetpointLeft(double setpoint)
+    {
+        _climbSetpointLeft = MathUtil.clamp(setpoint, _climbMinExtension, _climbMaxExtension);
+    }
+    
+    public void setSetpointRight(double setpoint)
+    {
+        _climbSetpointRight = MathUtil.clamp(setpoint, _climbMinExtension, _climbMaxExtension);
+    }
+
+    public void setVoltageLeft(double volts)
+    {
+        _climbSetpointLeft = null;
+        
+        _io.setLockStateLeft(Math.abs(volts) <= 0.12, _inputs); // TODO: Change number to constants
+        
+        _io.setVoltageLeft(volts);
+    }
```

Comment:
It's possible we might find ourselves in a position where we're using the desired height to automatically adjust the height of the arms using the algorithm, but then we directly set the voltage of only one side of the climber.

If this happens, the desired height setpoint will be cleared and the control algorithm will stop. The side that we set the voltage of will function correctly, but the other side will continue to move according to the last command it got from the algorithm.

To counteract this, we need to check if we're disabling the algorithm when setting the voltage directly. If we are, we need to stop the other side. I'm envisioning something like this:

```java
public void setVoltageLeft(double volts)
{
    if (_desiredHeight != null)
    {
        _desiredHeight = null;
        _io.setVoltageRight(0);
    }

    _io.setVoltageLeft(volts);
}
```